### PR TITLE
Plane: add As5600 angle-of-attack sensor

### PR DESCRIPTION
**Pull request by colejmero**
Branch would add library for AS5600 sensor to allow reading of angles.

### CODE REVIEW

File: ArduPlane/ArduPlane.cpp [hunk 1]
```diff
@@ -46,7 +46,8 @@ const AP_Scheduler::Task Plane::scheduler_tasks[] = {
     SCHED_TASK(navigate,               10,    150),
     SCHED_TASK(update_compass,         10,    200),
     SCHED_TASK(read_airspeed,          10,    100),
-    SCHED_TASK(update_alt,             10,    200),
+    SCHED_TASK(read_aoa,               10,    200),
+    SCHED_TASK(update_alt,             10,    200),//This line added by Cole
```
Comment (peterbarker) [inline ArduPlane/ArduPlane.cpp:50]: No it wasn't :-)

File: ArduPlane/ArduPlane.cpp [hunk 2]
```diff
@@ -193,6 +194,14 @@ void Plane::update_compass(void)
     }
 }
 
+/*
+ * Read and log Angle of Attack                    This function is code added by Cole
+ */
+
+void Plane::read_aoa(void){
```
Comment (peterbarker) [inline ArduPlane/ArduPlane.cpp:201]: You can use `SCHED_TASK_CLASS` instead of this wrapper

File: ArduPlane/Plane.h [hunk 3]
```diff
@@ -996,6 +999,7 @@ class Plane : public AP_Vehicle {
     void read_airspeed(void);
     void rpm_update(void);
     void accel_cal_update(void);
+    void read_aoa(void);                         //This line added by Cole
```
Comment (peterbarker) [inline ArduPlane/Plane.h:1002]: Let's lose the `This line added by Cole` throughout this patchset.  You probably know enough `git` now to remove the crutch

File: ArduPlane/wscript [hunk 4]
```diff
@@ -1,40 +1,41 @@
-#!/usr/bin/env python
```
Comment (peterbarker) [inline ArduPlane/wscript:1]: You appear to have changed the line endings in this file, making review very difficult.  Please change the line endings back.

File: libraries/AP_AS5600_AOA/AS5600_AOA.cpp [hunk 5]
```diff
@@ -0,0 +1,266 @@
+/****************************************************
+ * AMS 5600 Angle of Attack Class for Ardupilot platform
+ * Author: Cole Mero
+ * Date: 15 Dec 2014
```
Comment (peterbarker) [inline libraries/AP_AS5600_AOA/AS5600_AOA.cpp:4]: This doesn't look right.

File: libraries/AP_AS5600_AOA/AS5600_AOA.cpp [hunk 6]
```diff
@@ -0,0 +1,266 @@
+/****************************************************
+ * AMS 5600 Angle of Attack Class for Ardupilot platform
+ * Author: Cole Mero
+ * Date: 15 Dec 2014
+ * File: AS5600_AOA.cpp
+ * Version 1.00
+ * www.ams.com
```
Comment (peterbarker) [inline libraries/AP_AS5600_AOA/AS5600_AOA.cpp:7]: Is this right?

Do you need a "based on" pointing to the original source here?

File: libraries/AP_AS5600_AOA/AS5600_AOA.cpp [hunk 7]
```diff
@@ -0,0 +1,266 @@
...
+
+/****************************************************
+ * Method: AS5600_AOA
+ * In: none
+ * Out: none
+ * Description: constructor for AMS 5600
+***************************************************/
+ AS5600_AOA::AS5600_AOA()
+{
+
```
Comment (peterbarker) [inline libraries/AP_AS5600_AOA/AS5600_AOA.cpp:30]: Extra whitespace.

File: libraries/AP_AS5600_AOA/AS5600_AOA.cpp [hunk 8]
```diff
@@ -0,0 +1,266 @@
...
+    bus = 1; //Sets the bus number for the device, unclear what this number should be, trial and error to make it work
+    address = 0x36; //This is the I2C address for the device, it is set by the manufacturer
+
+   /*load register values*/
+
+  _zmco = 0x00;
```
Comment (peterbarker) [inline libraries/AP_AS5600_AOA/AS5600_AOA.cpp:36]:
```suggestion
  static const uint8_t REG_ZMCO = 0x00;
```

throughout here.

Also, these should be moved down to `private:` section in the header, or to the top of the `cpp` file.

File: libraries/AP_AS5600_AOA/AS5600_AOA.cpp [hunk 9]
```diff
@@ -0,0 +1,266 @@
...
+  _stat = 0x0b;
+  _agc = 0x1a;
+  _mag_hi = 0x1b;
+  _mag_lo = 0x1c;
+  _burn = 0xff;
+
+}
+
+
+void AS5600_AOA::init(){
```
Comment (peterbarker) [inline libraries/AP_AS5600_AOA/AS5600_AOA.cpp:58]:
```suggestion
void AS5600_AOA::init()
{
```

File: libraries/AP_AS5600_AOA/AS5600_AOA.cpp [hunk 10]
```diff
@@ -0,0 +1,266 @@
...
+
+    dev = hal.i2c_mgr->get_device(bus, address);
```
Comment (peterbarker) [inline libraries/AP_AS5600_AOA/AS5600_AOA.cpp:60]: You must do a nullptr check here.

File: libraries/AP_AS5600_AOA/AS5600_AOA.cpp [hunk 11]
```diff
@@ -0,0 +1,266 @@
...
+
+    WITH_SEMAPHORE(dev->get_semaphore());
+    dev->set_speed(AP_HAL::Device::SPEED_LOW);
+    dev->set_retries(2);
+
```
Comment (peterbarker) [inline libraries/AP_AS5600_AOA/AS5600_AOA.cpp:65]:
```suggestion

```

File: libraries/AP_AS5600_AOA/AS5600_AOA.cpp [hunk 12]
```diff
@@ -0,0 +1,266 @@
...
+}
+
+/* mode = 0, output PWM, mode = 1 output analog (full range from 0% to 100% between GND and VDD*/
+void AS5600_AOA::setOutPut(unsigned char mode){
```
Comment (peterbarker) [inline libraries/AP_AS5600_AOA/AS5600_AOA.cpp:69]:
```suggestion
void AS5600_AOA::setOutPut(uint8_t mode)
{

```

Need to use stdint types throughout.  I haven't pointed all these places out.

File: libraries/AP_AS5600_AOA/AS5600_AOA.cpp [hunk 13]
```diff
@@ -0,0 +1,266 @@
...
+    unsigned char config_status;
```
Comment (peterbarker) [inline libraries/AP_AS5600_AOA/AS5600_AOA.cpp:70]: Need a nullptr check on dev here

File: libraries/AP_AS5600_AOA/AS5600_AOA.cpp [hunk 14]
```diff
@@ -0,0 +1,266 @@
...
+    config_status = readOneByte(_conf_lo);
```
Comment (peterbarker) [inline libraries/AP_AS5600_AOA/AS5600_AOA.cpp:71]:
```suggestion
    uint8_t config_status;
    if (!dev->read_registers(REG_CONF_LO, &config_status, 1)) {
        ...
    }
```

File: libraries/AP_AS5600_AOA/AS5600_AOA.cpp [hunk 15]
```diff
@@ -0,0 +1,266 @@
...
+    if(mode == 1){
+        config_status = config_status & 0xcf;
```
Comment (peterbarker) [inline libraries/AP_AS5600_AOA/AS5600_AOA.cpp:73]: Really need to explain what's going on here.

File: libraries/AP_AS5600_AOA/AS5600_AOA.cpp [hunk 16]
```diff
@@ -0,0 +1,266 @@
...
+
+    /* Note significant variance from the Arduino AMS_5600 library in this line.
+     *
+     * -> writeOneByte(_conf_lo, lowByte(config_status));
+     *  vs writeOneByte(_conf_lo, config_status); I have removed the lowByte() function
+     *  since it is part of the arduino.h library and I don't have access to it, however,
+     *  since it accesses the lowest byte in the variable, and an unsigned char such as
+     *  config_status only HAS one byte regardless, I think the end result should be identical?
+     */
+    writeOneByte(_conf_lo, config_status);
```
Comment (peterbarker) [inline libraries/AP_AS5600_AOA/AS5600_AOA.cpp:86]:
```suggestion
    if (!_dev->write_register(REG_CONF_LO, config_status)) {
        ....
    }

```

remove similar uses of readonebyte, readtwobytes and the write equivalents.

File: libraries/AP_AS5600_AOA/AS5600_AOA.cpp [hunk 17]
```diff
@@ -0,0 +1,266 @@
...
+ * Description: gets raw value of magnet position.
+ * start, end, and max angle settings do not apply
+*******************************************************/
+unsigned short AS5600_AOA::getRawAngle(void)
+{
+  unsigned short angleRaw = readTwoBytes(_raw_ang_hi, _raw_ang_lo);
+
+  unsigned short angleDegrees = angleRaw*0.087;
+
+  AP::logger().Write("AoAR", "Status, TimeUS, Angle", "iQH", int(bool(dev)), AP_HAL::micros64(), angleDegrees);
```
Comment (peterbarker) [inline libraries/AP_AS5600_AOA/AS5600_AOA.cpp:103]: By tradition only upper-case in the message names.
```suggestion
  AP::logger().Write("AOAR", "Status, TimeUS, Angle", "iQH", int(bool(dev)), AP_HAL::micros64(), angleDegrees);

```

Comment (peterbarker) [reply libraries/AP_AS5600_AOA/AS5600_AOA.cpp:103]: You also need a `TimeUS` in this message.

File: libraries/AP_AS5600_AOA/AS5600_AOA.cpp [hunk 18]
```diff
@@ -0,0 +1,266 @@
...
+}
+/*******************************************************
+ * Method: highByte
+ * In: Unsigned short
+ * Out: Highest or leftmost byte of the unsigned short
+ * Description: Takes in the unsigned short and returns
+ * the leftmost or highest bite
+*******************************************************/
+
+unsigned char AS5600_AOA::highByte(unsigned short short_in){
```
Comment (peterbarker) [inline libraries/AP_AS5600_AOA/AS5600_AOA.cpp:115]: There's a `HIGHBYTE` define available to accomplish the same thing.

Similarly `LOWBYTE`

File: libraries/AP_AS5600_AOA/AS5600_AOA.cpp [hunk 19]
```diff
@@ -0,0 +1,266 @@
...
+*******************************************************/
+/*unsigned short AS5600_AOA::setMaxAngle(unsigned short newMaxAngle)
+{
+  unsigned short retVal;
+  if(newMaxAngle == -1)
+  {
+    maxAngle = getRawAngle();
+  }
+  else
+    maxAngle = newMaxAngle;
```
Comment (peterbarker) [inline libraries/AP_AS5600_AOA/AS5600_AOA.cpp:153]:
```suggestion
  if(newMaxAngle == -1){
    maxAngle = getRawAngle();
  } else {
     maxAngle = newMaxAngle;
  }
```

This is commented out, so no more review on this for the time being.

File: libraries/AP_AS5600_AOA/AS5600_AOA.cpp [hunk 20]
```diff
@@ -0,0 +1,266 @@
...
+//}
+
+
+/*******************************************************
+ * Method: readOneByte
+ * In: register to read
+ * Out: data read from i2c
+ * Description: reads one byte register from i2c
+*******************************************************/
+int AS5600_AOA::readOneByte(uint8_t in_adr)
```
Comment (peterbarker) [inline libraries/AP_AS5600_AOA/AS5600_AOA.cpp:226]: Can kill these.

File: libraries/AP_AS5600_AOA/AS5600_AOA.h [hunk 21]
```diff
@@ -0,0 +1,93 @@
...
+    unsigned short setMaxAngle(unsigned short newMaxAngle = -1);
+    unsigned short getMaxAngle();
+
+    unsigned short setStartPosition(unsigned short startAngle = -1);
+    unsigned short getStartPosition();
+
+    unsigned short setEndPosition(unsigned short endAngle = -1);
+    unsigned short getEndPosition();
+
+    unsigned short getRawAngle(void);
```
Comment (peterbarker) [inline libraries/AP_AS5600_AOA/AS5600_AOA.h:32]: Generally fix up your scoping here - make them as tight as possible.

**Review by peterbarker — CHANGES_REQUESTED** [review]
`mav_0_1.parm`  has to be removed - and all the others you've added.  Generally avoid using "git commit -a" if that's how they got in.

I've skipped the commented out blocks.

Bit of work to be done here, but little to make it safe for flight.

**Comment by rmackay9** [comment]
In keeping with our regular conventions it would be good to prefix the commits with the vehicle, library or directory being modified.

For example the first commit should probably be renamed, "Plane: adds AS5600 library"

File: libraries/AP_AS5600_AOA/AS5600_AOA.cpp (hunk 17 above)
Comment (hendjoshsr71) [reply libraries/AP_AS5600_AOA/AS5600_AOA.cpp:103]: `TimeUS` is there just hiding in second position and should be moved to first

Comment (colejmero) [reply libraries/AP_AS5600_AOA/AS5600_AOA.cpp:103]: I'm a little confused, do I not currently have one? At least...it looks to me like I do.

Comment (peterbarker) [reply libraries/AP_AS5600_AOA/AS5600_AOA.cpp:103]: Ah - sorry, tradition makes it the first entry :-)

**Comment by IamPete1** [comment]
I don't want to feature creep you to much, but it would be great to have a generic backed for this sensor. AP_Encoder or similar. The has been some interest in using them for sailboat windvanes, wheel encoders on rover also jump to mind.

**Comment by peterbarker** [comment]
On Wed, 3 Mar 2021, Peter Hall wrote:

> I don't want to feature creep you to much, but it would be great to have a generic backed for this sensor. AP_Encoder or similar. The has been some
> interest in using them for sailboat windvanes, wheel encoders on rover also jump to mind.

This will probably end up in AP_AOA/AP_AOA_AS5600.h etc - is that near 
enough to what you had in mind?

... and "AP_Encoder"?!  Really? :-P

File: libraries/AP_AS5600_AOA/AS5600_AOA.cpp (hunk 15 above)
Comment (colejmero) [reply libraries/AP_AS5600_AOA/AS5600_AOA.cpp:73]: This was ported over from the Arduino library, I don't actually understand it myself yet, will have to do more research/investigating.

File: libraries/AP_AS5600_AOA/AS5600_AOA.cpp [hunk 22]
```diff
@@ -0,0 +1,177 @@
...
+bool AS5600_AOA::init()
+{
+    dev = hal.i2c_mgr->get_device(bus, address);
+
+    if (!dev){
+        return false;
+    }
+
+    WITH_SEMAPHORE(dev->get_semaphore());
+    dev->set_speed(AP_HAL::Device::SPEED_LOW);
```
Comment (peterbarker) [inline libraries/AP_AS5600_AOA/AS5600_AOA.cpp:44]: We should be looking for a `WHOAMI` register here to make sure we're talking to who we're supposed to be talking to.

**Review by peterbarker — CHANGES_REQUESTED** [review]
Remove stray files.

Lots of unimplemented methods and unused state in the object not commented on as I believe you wanted to flesh this stuff out.

**Comment by peterbarker** [comment]
Closing in favour of https://github.com/ArduPilot/ardupilot/pull/16397

**Pull request closed without merging**